Update docs for Model-centric UI (#873)
grouparoo/www.grouparoo.com: pages/docs/config/sources.mdx

---
title: Sources
date: "2022-01-12"
pullQuote: "Learn how to configure sources using Grouparoo's UI Config."
---

<Alert variant="warning">
  <p className="mb-0">
    Sources require a <a href="/docs/config/models">Model</a> and an{" "}
    <a href="/docs/config/apps">App</a>.
  </p>
</Alert>

> A Source is a definition for how you pull data from an App into Grouparoo. [See this doc](/docs/getting-started/product-concepts) to learn more about the core concepts in Grouparoo.

Reviewer: ```suggestion
A Source is a definition for how you pull data from an App into Grouparoo. [See this page](/docs/getting-started/product-concepts#source) to learn more about the core concepts in Grouparoo.
Write
```

A Model has two types of Sources:

> - **Primary Source**: This is the source that provides the bulk of the data to the Model. You can only have one per Model.

Reviewer: ```suggestion
- **Primary Source**: This is the source that provides the bulk of the data to the Model. You can only have one per Model. The Primary Source will provide the unique identifier for Records in this Model, like `user_id` or `email_address`.
```

- **Secondary Sources**: These sources supplement the data provided by the Primary Source.

### Step 1: Choose an App

To start, click on your Model in the left side navigation and click `Create Primary Source` (or `Add new Source` for Secondary Sources).

<ImageInBrowserFrame
  centered
  alt="UI Config: New Source"
  src="/images/docs/guides/config/source--01-add.png"
  width={1005}
  height={628}
/>

> Choose the app you'd like use to feed your source. The choices shown will be based on which apps you've already connected. If you don't see the App you want to use, you can [add a new App](/docs/config/apps).

Reviewer 2: ```suggestion
Choose the App you'd like to use to feed your source. The choices shown will be based on which Apps you've already connected. If you don't see the App you want to use, you can [add a new App](/docs/config/apps).
```

<ImageInBrowserFrame
  centered
  alt="UI Config: Choose App"
  src="/images/docs/guides/config/source--02-choose-app.png"
  width={1062}
  height={664}
/>

### Step 2: Configure Your Source

Once you've chosen the App, fill in the appropriate configuration values for your source. This will determine what information from the source gets imported into Grouparoo.

> You will also need to indicate what makes each record in your source unique. This is a unique identifier such a `user id`, `uid`, or `email`.

Reviewer 2: ```suggestion
You will also need to indicate what makes each Record in your Source unique. This must be a unique identifier such a `user id`, `uid`, or `email`.
```

> When defining the Model's Primary Source, you will only need to select the identifier and

Reviewer 2: ```suggestion
When defining the Model's Primary Source, you will only need to select the identifier.
```

> Grouparoo will generate a Property that will serve as the Primary Key for the Model where the source is located.

Reviewer 2: ```suggestion
Then, Grouparoo will generate a Property that will serve as the Primary Key for the Model where the Source is located.
```

In the example below, the `id` column of the `users` table will be used to identify each unique record.

<ImageInBrowserFrame
  centered
  alt="UI Config: Configure Source"
  src="/images/docs/guides/config/source--03-configure.png"
  width={1061}
  height={746}
/>

> When defining a Secondary Source, you can map an attribute in the Source back to an existing property that ties it back to the Primary Source.

Reviewer 2: ```suggestion
When defining a Secondary Source, you can map an attribute in the Source to an existing property that ties it back to the Primary Source.
```

> In the example below, the Pg Purchases source is being tied back to the Pg Users source via the `user_id` foreign key in the `purchases` table.

Reviewer 2: ```suggestion
In the example below, the Pg Purchases Source is being tied back to the Primary Source (Pg Users) via the `user_id` foreign key in the `purchases` table.
```

---
title: Sources
date: "2022-01-12"
pullQuote: "Learn how to configure sources using Grouparoo's UI Config."
---

<Alert variant="warning">
  <p className="mb-0">
    Sources require a <a href="/docs/config/models">Model</a> and an{" "}
    <a href="/docs/config/apps">App</a>.
  </p>
</Alert>

A Source is a definition for how you pull data from an App into Grouparoo. [See this page](/docs/getting-started/product-concepts#source) to learn more about the core concepts in Grouparoo.
Write

A Model has two types of Sources:

- **Primary Source**: This is the data source that provides the bulk of the data to your Model. You can only have one Primary Source per Model.
  The Primary Source will provide the unique identifier for each Record in the Model such as `user_id` or `email_address`.
- **Secondary Sources**: These sources supplement the data provided by the Primary Source.

See the [Product Concepts documentation](/docs/getting-started/product-concepts) to learn more.

### Step 1: Choose an App

> Select on your Model in the left side navigation and click the "Create Primary Source" (or "Add new Source" for Secondary Sources) button.

Reviewer 3: ```suggestion
Select your Model in the left side navigation and click the "Create Primary Source" (or "Add new Source" for Secondary Sources) button.
```

<ImageInBrowserFrame
  centered
  alt="UI Config: Choose App"
  src="/images/docs/guides/config/source--02-choose-app.png"
  width={1062}
  height={664}
/>

### Step 2: Configure Your Source

Once you've chosen the App, fill in the appropriate configuration values for your source. This will determine what information from the source gets imported into Grouparoo.

You will also need to indicate what makes each Record in your Source unique. This must be a unique identifier such a `user id`, `uid`, or `email`.

When defining the Model's Primary Source, you will only need to select the identifier.
Then, Grouparoo will generate a Property that will serve as the Primary Key for the Model where the Source is located.

In the example below, the `id` column of the `users` table will be used to identify each unique record.

<ImageInBrowserFrame
  centered
  alt="UI Config: Configure Source"
  src="/images/docs/guides/config/source--03-configure.png"
  width={1061}
  height={746}
/>

When defining a Secondary Source, you can map an attribute in the Source to an existing property that ties it back to the Primary Source.

> In the example below, the Pg Purchases Source is being tied back to the Primary Source (Pg Users) via the `user_id` foreign key in the `purchases` table.

Reviewer 3: ```suggestion
In the example below, the "Pg Purchases" Source is being tied back to the Primary Source (Pg Users) via the `user_id` foreign key in the `purchases` table.
```

/>

Choose the App you'd like to use to feed your Source. The choices shown will be based on which Apps you've already connected. If you don't see the App you need, you can [add a new App](/docs/config/apps).

<ImageInBrowserFrame
  centered
  alt="UI Config: Choose App"
  src="/images/docs/guides/config/source--02-choose-app.png"
  width={1062}
  height={664}
/>

### Step 2: Configure Your Source

Once you've chosen the App, fill in the appropriate configuration values for your Source. This will determine what information from the Source gets imported into Grouparoo.

You will also need to indicate what makes each Record in your Source unique. This must be a unique identifier such a `user id`, `uid`, or `email`.

When defining the Model's Primary Source, you will only need to select the identifier.
Then, Grouparoo will generate a Property that will serve as the Primary Key for the Model where the Source is located.

In the example below, the `id` column of the `users` table will be used to identify each unique record.

<ImageInBrowserFrame
  centered
  alt="UI Config: Configure Source"
  src="/images/docs/guides/config/source--03-configure.png"
  width={1061}
  height={746}
/>

> 

Reviewer: Maybe there should be sections
```
#### Primary Source
#### Secondary Source
```
